### PR TITLE
Updated readme and strategy with doc

### PR DESCRIPTION
**Pull request by Bigood**
Allowed string error to fail the strategy without crashing it, type of createProvider checked, documentation added and example added in the readme

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,6 +53,39 @@ var strategy = new LTIStrategy({
...
+          var consumer = new lti.Provider(consumer_db.oauth_consumer_key, consumer_db.oauth_consumer_secret);
+          return done(null, consumer);
+        }
+        else {
+          // String error, will fail the strategy (and not crash it)
+          return done("not_authorized");
+        }
+	    }
+	  );
+	}
```
Comment (lzhang) [inline README.md]: Can you fix the whitespace here?

File: lib/strategy.js [hunk 2]
```diff
@@ -2,6 +2,34 @@ var lti = require('ims-lti');
...
+ *      and the process is just intended to stop.
+ *
+ *      Use either this function or the hardcoded key / secret. 
+ *      This one gets priority over the hardcoded key / secret.
+ *
+ *      @param {Function} createProvider 
+ *        @param {Object} req
+ *        @param {Function} callback
+ *           @param {Object || String} err
+ *           @param {Object} provider
```
Comment (lzhang) [inline lib/strategy.js]: More whitespace strangeness.

File: lib/strategy.js [hunk 3]
```diff
@@ -10,9 +38,18 @@ function Strategy(options, verify) {
...
+  if (typeof(options.createProvider) === "function") {
+    // This will get called when the request authenticates
+    // And the LTI Provider will be defined at this time, for each request
     this._createProvider = options.createProvider;
-  } else {
+  } 
+  // Else, we assume we're in a development environment, 
+  // thus there's only one test consumer
+  else {
+    // Instanciate the provider from there, with hardcoded key & secret
```
Comment (lzhang) [inline lib/strategy.js]: Spelling "Instantiate".

File: lib/strategy.js [hunk 4]
```diff
@@ -10,9 +38,18 @@ function Strategy(options, verify) {
   passport.Strategy.call(this);
   this.name = 'lti';
   this._passReqToCallback = options.passReqToCallback;
-  if (options.createProvider) {
+
+  // If it's intended to check tool consumer's identity from database
+  // Dev will pass a function that handle it
+  if (typeof(options.createProvider) === "function") {
```
Comment (lzhang) [inline lib/strategy.js]: `typeof` is a keyword, not a function.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof

File: lib/strategy.js [hunk 5]
```diff
@@ -37,9 +74,20 @@ Strategy.prototype.authenticate = function(req, options) {
...
+  // Check consumer's identity with strategy's predefined database access function
   self._createProvider(req, function(err, provider) {
-    if (err) return self.error(err);
-
+    // Handle errors
+    if (err) {
+      // Non standard string error
+      if (typeof(err) == "string") {
+        // Fail the strategy with the error message
+        return self.fail(err);
```
Comment (lzhang) [inline lib/strategy.js]: Thanks, this seems like a big oversight on my part.

File: lib/strategy.js [hunk 6]
```diff
@@ -37,9 +74,20 @@ Strategy.prototype.authenticate = function(req, options) {
...
   }
 
+  // Check consumer's identity with strategy's predefined database access function
   self._createProvider(req, function(err, provider) {
-    if (err) return self.error(err);
-
+    // Handle errors
+    if (err) {
+      // Non standard string error
+      if (typeof(err) == "string") {
```
Comment (lzhang) [inline lib/strategy.js]: Ditto `typeof`.

File: lib/strategy.js (hunk 4 above)
Comment (Bigood) [inline lib/strategy.js]: True, thanks for pointing that out.

**Pull request merged**